addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/handler/XiaomiActorGatewayHandler.java

```diff
@@ -0,0 +1,222 @@
+                    int kelvin = 48 * colorTemperature.intValue() + 1700;
+                    int color = ColorUtil.getRGBFromK(kelvin);
+                    writeBridgeLightColor(color, getGatewayLightBrightness());
+                    updateState(CHANNEL_COLOR,
+                            HSBType.fromRGB((color / 256 / 256) & 0xff, (color / 256) & 0xff, color & 0xff));
+                } else {
+                    logger.error("Can't handle command {} on channel {}", command, channelUID);
+                }
+                break;
+            case CHANNEL_GATEWAY_SOUND:
+                if (command instanceof DecimalType) {
+                    Item volumeItem = getItemInChannel(CHANNEL_GATEWAY_VOLUME);
+                    State state;
+                    if (volumeItem == null) {
+                        state = null;
+                        logger.debug("There was no Item found for soundVolume, default 50% is used");
+                    } else {
+                        state = volumeItem.getState();
+                    }
+                    int volume = (state instanceof DecimalType && state != null) ? ((DecimalType) state).intValue()
+                            : 50;
+                    writeBridgeRingtone(((DecimalType) command).intValue(), volume);
+                    updateState(CHANNEL_GATEWAY_SOUND_SWITCH, OnOffType.ON);
+                } else {
+                    logger.error("Can't handle command {} on channel {}", command, channelUID);
+                }
+                break;
+            case CHANNEL_GATEWAY_SOUND_SWITCH:
+                if (command instanceof OnOffType) {
+                    if (((OnOffType) command) == OnOffType.OFF) {
+                        stopRingtone();
+                    }
+                } else {
+                    logger.error("Can't handle command {} on channel {}", command, channelUID);
+                }
+                break;
+            case CHANNEL_GATEWAY_VOLUME:
+                // nothing to do, just suppress error
+                break;
+            default:
+                logger.error("Can't handle command {} on channel {}", command, channelUID);
+                break;
+        }
+    }
+
+    @Override
+    void parseReport(JsonObject data) {
+        parseHeartbeat(data);
+    }
+
+    @Override
+    void parseHeartbeat(JsonObject data) {
+        if (data.has("rgb")) {
+            long rgb = data.get("rgb").getAsLong();
+            updateState(CHANNEL_BRIGHTNESS, new PercentType((int) (((rgb >> 24) & 0xff))));
+            updateState(CHANNEL_COLOR,
+                    HSBType.fromRGB((int) (rgb >> 16) & 0xff, (int) (rgb >> 8) & 0xff, (int) rgb & 0xff));
+        }
+        if (data.has("illumination")) {
+            int illu = data.get("illumination").getAsInt();
+            updateState(CHANNEL_ILLUMINATION, new DecimalType(illu));
+        }
+    }
+
+    @Override
+    void parseReadAck(JsonObject data) {
+        parseHeartbeat(data);
```

Comment:
If all child classes have something like this you could consider the following pattern in the parent class:

```java
void parseReadAck(JsonObject data) {
    parseDefault(data); // by default assume all messages are the same
}

...

abstract void parseDefault(JsonObject data);
```

You would then only implement the `parseDefault` and any method that is not parsed in the default way. The current implementation looks strange because sometime you interpret all as an heartbeat and in others as something else.